components/VM_Arm/src/modules/virtio_con.c

```diff
@@ -22,61 +42,160 @@
 #include <platsupport/serial.h>
 #include <virtioarm/virtio_console.h>
 
-static virtio_con_t *virtio_con = NULL;
-extern void *serial_getchar_buf;
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
 
-//this matches the size of the buffer in serial server
-#define BUFSIZE 4088
+/* Number of ports is the number of crossvm connections plus hvc0 */
+#define NUM_PORTS (ARRAY_SIZE(serial_layout) + 1)
 
-char txbuf[BUFSIZE];
+static virtio_con_t *virtio_con = NULL;
 extern vmm_pci_space_t *pci;
 extern vmm_io_port_list_t *io_ports;
 
-struct {
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE (0x1000 - 2 * sizeof(uint32_t))
+
+/**
+ * This struct is a ring buffer that occupies exactly one page, it should represent
+ * the data in the shmem region (that is shared between two nodes of a virtio console
+ * connection.
+ *
+ * The access can be lock-free since an instance of this struct can only be either the
+ * receive buffer or the sent buffer of a virtio console node, which guarantees that only
+ * one end of the ring buffer struct can be modified in a node.
+
+ * Way of indexing: `head` and `tail` should not increase beyond (BUFSIZE - 1), as
+ * they repersent the real index in `buf`.
+ */
+typedef struct serial_shmem {
     uint32_t head;
     uint32_t tail;
     char buf[BUFSIZE];
-} extern volatile *batch_buf;
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* defined in virtio_console_emul.c */
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buffer, uint32_t head, uint32_t tail) WEAK;
```

Comment:
Can't we move this prototype definition to a header file? I also wonder, what does this do practically? It make some function implemented somewhere else weak, that seems a bit odd. The weak should be where the implementation is?

Reply:
based on previous implementations of virtio backends, instead of calling the function `virtio_console_putchar` directly from this file, there should be an "emul driver" structure that contains a pointer to this function that got passed from the emul layer to this layer.
I'll make it so unless anyone has objections.